Add Morgengrauen connection option

Kebap opened the pull request:
#### Brief overview of PR changes/additions
Adds default connection info for Morgengrauen.

#### Motivation for adding to Mudlet
Add to the list of games offering a good experience in Mudlet and make it easier for players to find MG

#### Other info (issues closed, discussion etc)
Currently still draft status, some information and code missing, using Github for faster compilation

Working to make sure everything in https://wiki.mudlet.org/w/Listing_Your_MUD is kosher :)

Looking forward to continue improving the UI even more in the future.

As the game doesn't want to push packages via GMCP automatically, in lieu of other players using other clients maybe, we include a small loader script to download & install the UI to deliver an out of the box experience.

Clarification: The game will push packages via GMCP alright (if you were to type an ingame command), and has also already implemented GMCP data in the past with details for character and room, etc.

To test using a fresh installation: https://wiki.mudlet.org/w/Manual:Testing#Simulate_a_fresh_install

Kebap committed:
Try listing a new game

Kebap committed:
Try adding discord rich support

Kebap commented:
/create links

Kebap commented:
Whatever, here are the downloads: 
https://make.mudlet.org/snapshots/?platform=all&source=pull-request
Look for this PR number 6499

Building a new version takes some time.. Linux: 12min, MacOS: 15min, Windows: 23min

Kebap committed:
try adding a bootstrap package

Kebap committed:
4711 is insecure, 4712 would be secure port

Kebap committed:
rename for consistency with cf-loader.xml

Kebap committed:
Remove third-party texts

keneanung commented:
/create links

Kebap committed:
Actually load the bootstrap package

SlySven commented on src/TGameDetails.h:
```diff
@@ -535,6 +543,24 @@ class TGameDetails
+             4711,
+             false,
+             qsl("<a href='http://mg.mud.de/'>mg.mud.de</a>"),
+             qsl(":/icons/morgengrauen_480x120.png"),
+             /* English translation, provided by Game:
+ *           "Welcome to Morgengrauen, the largest German-language "
+ *           "MUD, which has been continuously developed since 1992."
+ *           "\n\n"
+ *           "Play now or never!"
+ */
+             qsl("Willkommen im Morgengrauen, dem größten deutschsprachigen "
```
```suggestion
             qsl(":/icons/morgengrauen_480x120.png"),
             /* English translation, provided by Game:
              * "Welcome to Morgengrauen, the largest German-language "
              *  "MUD, which has been continuously developed since 1992."
              * "\n\n"
              * "Play now or never!"
              */
             qsl("Willkommen im Morgengrauen, dem größten deutschsprachigen "
```

SlySven commented on src/TGameDetails.h:
```diff
@@ -62,7 +62,15 @@ class TGameDetails
 
     // clang-format off
-    // games are to be added here in alphabetical order
+    //
+    // Games are to be added here in alphabetical order and this format:
+    // - Name
+    // - Address to connect to
+    // - Port to connect on
+    // - Secure connection possible?
+    // - Website or other URLs
+    // - Path to the profile icon
+    // - Text to use for description
```
Not needed - I have already documented the first entry - if you insist that it needs to be done here, like this, then please remove corresponding comments from that first entry.
```suggestion
    // games are to be added here in alphabetical order
```

Kebap replied:
No hard feelings on this one, but I agree it maybe should not be documented twice. 
Probably a matter of taste, but I found the inline-documentation less readable albeit more informative.
Leaving this open for a moment for others to review and maybe add their opinion as well.
Then again, this probably should not be part of adding a game at all, but a PR of its own.

Kebap committed:
Include mg-loader.xml in mudlet.pro

SlySven reviewed (commented):
I am not sure about including "[ ERROR ]  -" et al. type messages in the Lua subsystem (as a script) given that those are normally included in GUI type messages that are to be translated for the end-user's choice of locale.

As such, maybe we need a way to include translations for at least the locales we support for such UI messages to be accessed in a way that can work alongside the `QObject::tr(....)` system - IIRC we do do something like that for native exit directions but I have not actually looked at that.

The icon for the MUD has the wrong aspect ratio - it should be 3:1 not 4:1 - so either 360x120  or 480x160 ...

Kebap committed:
Straighten bootstrap package some more

Kebap commented:
> I am not sure about including "[ ERROR ] -" et al. type messages in the Lua subsystem (as a script) given that those are normally included in GUI type messages that are to be translated for the end-user's choice of locale.
>
> As such, maybe we need a way to include translations for at least the locales we support for such UI messages to be accessed in a way that can work alongside the QObject::tr(....) system - IIRC we do do something like that for native exit directions but I have not actually looked at that.
>
These messages were inspired from cf-loader.xml, where they serve the very same purpose, but use arbitrary colors.
I aligned the colors and layout of the text to resemble something the player could know and recognize more easily.

> The icon for the MUD has the wrong aspect ratio - it should be 3:1 not 4:1 - so either 360x120 or 480x160 ...
>
I did not find this true. Please compare:
```
3klogo.png		130	40	13:4
3slogo.png		717	365	wtf
120x30RoDLogo.png	120	30	4:1
aardwolf_mud.png	120	27	4:0.9
achaea_120_30.png	120	30	4:1
aetolia_120_30.png	120	30	4:1
avalon.png		120	27	4:0.9
batmud_mud.png		120	27	4:0.9
carrionfields.png	360	90	4:1
...
morgengrauen_480x120	480	120	4:1
```

Edited by @SlySven a second time to restore original comment from @Kebap - sorry about that.

Kebap committed:
Update src/TGameDetails.h

Align whitespace like SlySven suggested

Co-authored-by: Stephen Lyons <slysven@virginmedia.com>

Kebap commented:
@SlySven DId you by chance edit my comment instead of quoting / replying?

Kebap committed:
Undo comments on game structure - out of scope here

Co-authored-by: Stephen Lyons <slysven@virginmedia.com>

Kebap committed:
drop discord support for now

Discord will be handled via Lua scripts for now until the game server cooperates better

SlySven commented:
> @SlySven DId you by chance edit my comment instead of quoting / replying?

:hankey: Oh, blooming 'eck, yeah it looks like I did, I meant to quote you and reply and it looks like I edited your post instead. I've tried to put the comment back to your original, I'll put response here:


>>I am not sure about including "[ ERROR ] -" et al. type messages in the Lua subsystem (as a script) given that those are normally included in GUI type messages that are to be translated for the end-user's choice of locale.
>>
>>As such, maybe we need a way to include translations for at least the locales we support for such UI messages to be accessed in a way that can work alongside the QObject::tr(....) system - IIRC we do do something like that for native exit directions but I have not actually looked at that.
>>
>These messages were inspired from cf-loader.xml, where they serve the very same purpose, but use arbitrary colors.
>I aligned the colors and layout of the text to resemble something the player could know and recognize more easily.
>
>>The icon for the MUD has the wrong aspect ratio - it should be 3:1 not 4:1 - so either 360x120 or 480x160 ...
>>
>I did not find this true. Please compare:
```
3klogo.png		130	40	13:4
3slogo.png		717	365	wtf
120x30RoDLogo.png	120	30	4:1
aardwolf_mud.png	120	27	4:0.9
achaea_120_30.png	120	30	4:1
aetolia_120_30.png	120	30	4:1
avalon.png		120	27	4:0.9
batmud_mud.png		120	27	4:0.9
carrionfields.png	360	90	4:1
...
morgengrauen_480x120	480	120	4:1
```
:man_facepalming: forgive me - I think there must have been too much blood in my caffeine stream - you are correct and I got that wrong - so please ignore that complaint.

Kebap committed:
Make sure bootstrap package gets removed

Kebap committed:
Update image

Kebap committed:
Merge branch 'development' into morgengrauen

Kebap committed:
Merge branch 'development' into morgengrauen

SlySven commented:
I've just undone what **mudlet-machine-account** did (I think we've "fixed" it since it last acted on this PR) and put it back to how I think @vadi2 put it... :crossed_fingers:

Kebap commented:
I found this OK to be included in 4.17 (if it is done by then) because it's no new functionality in Mudlet itself.

Kebap committed:
Merge branch 'development' into morgengrauen

vadi2 commented:
This is still draft, which signals that it's not ready for review - is that the intention?

Kebap commented:
> This is still draft, which signals that it's not ready for review - is that the intention?

Well while this PR is unlikely to change much, it will download the game package, which is still in final developments, so yeah.

That said, if you do review already, and find anything needs adjustment in this very PR here, I would be interested to know.
So maybe I should lift draft status after all? 🤔 Not sure.

Kebap committed:
Merge branch 'development' into morgengrauen

vadi2 commented:
There's a lot of other work to be done so let me know when this is ready to be reviewed by hitting the `Ready for review` button. If anyone else is willing to give it a go in the meantime, that is welcome too.

Kebap committed:
Merge branch 'development' into morgengrauen

SlySven reviewed (commented):
Do you still need to keep M-G out of the list in `./src/discord.cpp` as you did in 7c2e38a57d1eb225eae885d9adf8a94559d79529 ?

Also I am not sure that the flag in the corner of the game icon is big enough to signal that this is a German MUD (and the Italian one for *Clessidra* isn't much better) TBH I think that it needs to be twice as big in both directions (so four times the area) to show up properly on, at least, one of my 1920x1080 monitors...
![image](https://user-images.githubusercontent.com/6163092/221603795-14bfde9f-e0da-4726-9c0a-90f4320ee6f3.png)

Otherwise this looks good to me. :+1:

Kebap commented:
> Do you still need to keep M-G out of the list in `./src/discord.cpp` as you did in [7c2e38a](https://github.com/Mudlet/Mudlet/commit/7c2e38a57d1eb225eae885d9adf8a94559d79529) ?

Thanks for reminding this fix. I tested that before and it didn't work out, so I removed it for now to keep scope manageable. Will require more admin attention down the line. Currently discord information will be handled by and originate in Lua land only, no interaction from game there, yet. 

> Also I am not sure that the flag in the corner of the game icon is big enough to signal that this is a German MUD (and the Italian one for _Clessidra_ isn't much better) 

Yep, that's where I aligned on. We may need to discuss this further in a new issue, because there are other international games without any flag at all, so this makes Clessidra and MorgenGrauen now best in class. 

> Otherwise this looks good to me. 👍

Thanks!

SlySven commented:
> > Do you still need to keep M-G out of the list in `./src/discord.cpp` as you did in [7c2e38a](https://github.com/Mudlet/Mudlet/commit/7c2e38a57d1eb225eae885d9adf8a94559d79529) ?
> 
> Thanks for reminding this fix. I tested that before and it didn't work out, so I removed it for now to keep scope manageable. Will require more admin attention down the line. Currently discord information will be handled by and originate in Lua land only, no interaction from game there, yet.

IIRC A mention there is because @vadi2 (as the Discord Guild owner) has uploaded an icon that is stored there (under the lower-case ASCII alphabet name of `morgengrauen`) and can be used when (by default) Mudlet's own Discord Application ID is being used - so that the Rich Presence will say "Playing Mudlet" in the Discord User List  and "Mudlet" on the top line of text and "Playing morgengrauen" on the second line in the absence and the game icon as the large icon when the Rich Presence is viewed - all in the ***absence*** of a dedicated MUD Server Application ID being set...

Kebap committed:
Merge branch 'Mudlet:development' into morgengrauen

Kebap committed:
whitespace & colors

colors copied from `void cTelnet::postMessage(QString msg)`

Kebap commented:
> ... can be used when (by default) Mudlet's own Discord Application ID is being used - so that the Rich Presence will say "Playing Mudlet" in the Discord User List and "Mudlet" on the top line of text and "Playing morgengrauen" on the second line in the absence and the game icon as the large icon when the Rich Presence is viewed - all in the _**absence**_ of a dedicated MUD Server Application ID being set...

This in not the case. There will be a dedicated ID used for this game, and the Rich Presence will be configured by Lua itself.

vadi2 commented:
Cleared the 4.17 milestone as this isn't intended to be a blocker to releasing - we can review & merge this whenever it is ready.

SlySven committed:
Fix: try and solve crashing when saving profile asynchronously

It looks as though this is due to a dangling reference because we never
take a copy of the file name when we start the `QConcurrent::run(...)`.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Fix_tryAndSolveCrashingWhenSavingProfileAsynchronously

Kebap committed:
Merge branch 'development' into morgengrauen

Kebap committed:
Merge remote-tracking branch 'SlySven/Fix_tryAndSolveCrashingWhenSavingProfileAsynchronously' into morgengrauen

Kebap committed:
Remove whitespace from decho parameters

Kebap committed:
Fix newline characters in Lua long strings

Kebap commented:
This is good to review now, and include in next development merge as you see fit.

To test the login, please use the following details:
- gast
- w

![image](https://user-images.githubusercontent.com/117238/229329645-b5e8dce2-f2bd-495e-a865-b5cd13d9d5e7.png)

Kebap committed:
Improve game description

vadi2 reviewed (approved):
Looks great, thanks!

By the way you can adjust the generic mapper not to be installed in `mudlet::setupPreInstallPackages`. I'll merge for now though as you have a workable solution already
